keps/sig-network/4444-service-traffic-distribution/README.md

```diff
@@ -311,14 +324,29 @@ the value configured for `trafficDistribution`
 * This leverages existing implementation, requiring no major changes.
 
 #### `PreferClose`
-* **Meaning:** Attempts to route traffic to endpoints within the same zone as
-  the client. If no endpoints are available within the zone, traffic would be
-  routed to other zones.
+* **Meaning:**
+  * Attempts to route traffic to endpoints within the same zone as the client. If no
+    endpoints are available within the zone, traffic would be routed to other zones.
+  * If the Service has endpoints that are part of a DaemonSet deployed to all nodes,
+    then it attempts to route traffic to endpoints on the same node as the client. If
+    no endpoints are available on the same node, traffic would be routed to other
+    nodes (though still preferring the same zone, if relevant).
+
 * This preference will be implemented by the use of Hints within EndpointSlices.
-* We already use Hints to implement `service.kubernetes.io/topology-mode: Auto`
-  In a similar manner, the EndpointSlice controller will now also populate hints
-  for `trafficDistribution: PreferClose` -- although in this case, the zone hint will
-  match the endpoint of the zone itself.
+  * We already use `Hints.ForZones` to implement
+    `service.kubernetes.io/topology-mode: Auto` In a similar manner, the
+    EndpointSlice controller will now also populate zone hints for
+    `trafficDistribution: PreferClose` -- although in this case, the zone hint will
+    match the endpoint of the zone itself.
+  * For the same-node preference, the EndpointSlice controller will set a new
+    `Hints.ForSameNode` field (to `true`) to indicate to the service proxy that
+    clients on the same node as this endpoint should prefer it. This will be set
+    when: 
+      * the Service has `PreferClose` traffic distribution, and
+      * the Service has endpoints on all or all-but-one of the Ready nodes, and
```

Comment:
This comment may be similar to what Antonio is saying below. 
If a DaemonSet is configured to for `updateStrategy.rollingUpdate.maxUnavailable` of > 1, this won't match.

Also, a use case we have is that we run a statsd DaemonSet on all nodes, except for a handful of nodes that are short-lived and have a very specific task, and running the DaemonSet there costs too much.

Reply:
And would you want prefer-same-node semantics for that Service?

There is definitely a tradeoff if we want to try to "auto-detect" this.

Maybe I should try to figure out [the `TopologySpreadConstraints` version of this](https://github.com/kubernetes/enhancements/pull/3293#issuecomment-1317254954)...

Reply:
> And would you want prefer-same-node semantics for that Service?

Definitely! 
Since we are running the DaemonSet on most of our nodes, we wouldn't worry about overloading a specific node.

What is the thought behind having the safe guards described here? Is the worry that a single node could be overloaded with traffic, and may be the correct choice would have been to send traffic to a different Pod (at the cost of going over the network).

I may be missing something here (context: I've never used the topology hints feature, but would really like to, so may be I haven't encountered its downsides yet), but the name "PreferClose" makes me assume that traffic would be sent to a pod on the same node if available, if it can't go to a local pod, then to a pod on the same zone, if that fails, then a random pod in the cluster.

Reply:
The issue is that we don't know anything about the _clients_ of the service, and we can't just assume that they are distributed in a way that ensures prefer-same-node will result in good behavior. If, for some reason, there are endpoints on Nodes A, B, and C, but most of the clients are on Node A, then prefer-same-node would result in the Node A endpoint getting overloaded while the Node B and C endpoints were idle. (If the nodes are heterogeneous in some way that affects scheduling this might even be a predictable outcome.)

If I was attaching the prefer-same-node semantics to a _new_ traffic distribution value then this wouldn't be a problem because we could just say "don't use `trafficDistribution: PreferSameNode` if you don't want that". But since there are already people using `PreferClose`, we have to be careful to not break them.

Reply:
This restriction seems like it's trying to provide the same kind of guardrails I tried to provide with the initial iteration of "topology aware hints". At the time, my working theory was that adding guardrails would help avoid overloaded endpoints, which was at least theoretically true, but it came at a very significant cost of confusing users.

I think this would have the same risks:

1. Do control plane nodes count?
2. Why are n-1 nodes acceptable but not n-2? This has a wildly different if you have 3 nodes or 3000 nodes in your cluster.

I'd strongly prefer following the TrafficDistribution behavior we settled on for zones - if a healthy endpoint matches the "close" definition, we'll route to it, otherwise we spray traffic everywhere. If both zone and node hints were set, it would probably be reasonable for kube-proxy to try to find an endpoint with a matching node, and if not, try to find an endpoint with a matching zone.